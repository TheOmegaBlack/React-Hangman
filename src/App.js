import React, { useState, useEffect, useRef } from 'react';
import './App.scss';
import StartScene from './StartScene';
import GameScene from './GameScene';

const maxLives = 7;
const maxTime = 30;

function App() {
  const [newGame, setNewGame] = useState(true);
  const [secretWord, setSecretWord] = useState('');
  const [enteredLetters, setEnteredLetters] = useState({});
  const [lives, setLives] = useState(maxLives);
  const [time, setTime] = useState(maxTime);
  const [countdown, setCountdown] = useState(() => {});
  const [playerWon, setPlayerWon] = useState(null);

  const appRef = useRef(null);

  const handleKeyInput = (e, playerWon, newGame, secretWord) => {
    const { keyCode, key } = e;

    // Checks if it's a letter, otherwise returns;
    if (keyCode < 65 && keyCode > 90) {
      return;
    }

    // Checks if the game is still ongoing
    if (playerWon === null && !newGame) {
      // Sets the entered letter
      setEnteredLetters(prevState => {
        if (!prevState[key]) {
          return {
            ...prevState,
            [key]: true,
          };
        }
        return prevState;
      });

      // Check if the secret word includes the letter or reduces the player's lives by one
      if (!secretWord.includes(key)) {
        setLives(prevLives => prevLives - 1);
      }
    }
  };

  // Side effect that happens when newGame changes
  useEffect(() => {
    // Used to focus on the div so that the player doesn't have to click to enter
    if (!newGame) {
      appRef.current.focus();
    }
  }, [newGame]);

  // "Side effect" that happens when typing letters
  useEffect(() => {
    const allLetters =
      secretWord.length &&
      secretWord.split('').every((l, i, { length }) => {
        return i === 0 || i === length - 1 || Boolean(enteredLetters[l]);
      });

    if (allLetters) {
      setPlayerWon(true);
      clearInterval(countdown);
    } else {
      setTime(maxTime);
    }
  }, [enteredLetters, secretWord, countdown]);

  // This side effect is called when player lives change
  useEffect(() => {
    if (lives <= 0) {
      setPlayerWon(false);
      clearInterval(countdown);
    }
  }, [lives, countdown]);

  // This side effect changes with time
  useEffect(() => {
    if (time === 0) {
      setLives(prevLives => prevLives - 1);
      setTime(maxTime);
    }
  }, [time]);

  // Function called when starting a new game
  const handleNewGame = () => {
    setNewGame(false);
    setCountdown(setInterval(() => setTime(prev => prev - 1), 1000));
  };

  // Function called to reset the game and start again
  const resetAll = () => {
    setNewGame(true);
    setSecretWord('');
    setEnteredLetters({});
    setLives(maxLives);
    setTime(maxTime);
    setPlayerWon(null);
  };

  return (
    <div
      className="App"
      data-testid="app-root"
      tabIndex={1}
      onKeyDown={e => handleKeyInput(e, playerWon, newGame, secretWord)}
      ref={appRef}
    >
      <main className="Main">
        <span className="Main__title">Sacha's Hangman</span>
        {newGame ? (
          <StartScene
            setNewGame={handleNewGame}
            setSecretWord={setSecretWord}
          />
        ) : (
          <GameScene
            lives={lives}
            time={time}
            secretWord={secretWord}
            enteredLetters={enteredLetters}
            resetAll={resetAll}
            playerWon={playerWon}
          />
        )}
      </main>
    </div>
  );
}

export default App;
