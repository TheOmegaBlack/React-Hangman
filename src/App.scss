@font-face {
  font-family: 'EncodeSans';
  src: url('fonts/EncodeSansSemiExpanded.ttf')
}

* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.App {
  background-color: #581845;
  font-family: 'EncodeSans', monospace;
  height: 100vh;
  padding-top: 5vh;
  outline: none;
  text-align: center;
}

.Main {
  background-color: #900C3F;
  border-radius: 25px;
  color: #ff5733;
  height: 500px;
  margin: 50px auto 0;
  width: 980px;

  &__title {
    display: block;
    font-size: 50px;
    margin: 0 10px 20px 0;
    text-align: right;
  }

  &__text {
    display: block;
    font-size: 40px;
    margin-top: 30px;
  }

  &__input {
    border: 0;
    border-radius: 5px;
    display: block;
    height: 40px;
    margin: 20px auto;
    width: 70%;
  }

  &__button {
    background-color: #FFC300;
    border: 0;
    border-radius: 5px;
    cursor: pointer;
    height: 50px;
    margin-top: 20px;
    width: 200px;
  }

  &__error {
    color: red;
    display: block;
  }

  &__lives, &__time {
    display: block;
    font-size: 30px;
    margin: 10px 0 30px 20px;
    text-align: left;
  }

  &__letters {
    font-size: 70px;
  }
}
