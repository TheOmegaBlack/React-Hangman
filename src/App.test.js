import React from 'react';
import ReactDOM from 'react-dom';
import {
  render,
  fireEvent,
  getByLabelText,
  getByText,
  queryByText,
  getByTestId,
  getNodeText,
} from '@testing-library/react';
import App from './App';

const gameSentence = 'gamesentence';

describe('App', () => {
  const { container } = render(<App />);

  it('renders correctly', () => {
    expect(container).toMatchSnapshot();
  });

  describe('tests with input in StartScene', () => {
    const inputNode = getByLabelText(container, 'Insert the word', {
      selector: 'input',
    });

    const doneButton = getByText(container, 'Done');

    const errorString = 'The word should be 5 and 15 characters long';

    it('correctly handles no value', () => {
      const value = inputNode.value;

      fireEvent.change(inputNode, {
        target: { value: '' },
      });

      expect(inputNode.value).toEqual(value);
    });

    it('cannot be sumitted when entering less than 5 characters', () => {
      fireEvent.change(inputNode, {
        target: { value: 'word' },
      });

      fireEvent.keyPress(inputNode, { key: 'Enter', code: 13, charCode: 13 });

      const errorSpan = getByText(container, errorString);

      expect(errorSpan).toBeTruthy();
      expect(queryByText(container, 'Done')).toBeTruthy();
    });

    it('shows errors if trying to enter more than 15 characters', () => {
      fireEvent.change(inputNode, {
        target: { value: 'thisisaveryverylongsentence' },
      });

      const errorSpan = getByText(container, errorString);

      expect(errorSpan).toBeTruthy();
    });

    it('does not show errors if entering between 0 and 15 characters', () => {
      fireEvent.change(inputNode, {
        target: { value: gameSentence },
      });

      const errorSpan = queryByText(container, errorString);

      fireEvent.click(doneButton);

      expect(errorSpan).toBeNull();
      expect(queryByText(container, 'Done')).toBeNull();
    });
  });

  describe('when in game scene', () => {
    const mainDiv = getByTestId(container, 'app-root');

    it('renders correctly after submitting a word', () => {
      expect(container).toMatchSnapshot();
    });

    it('reacts correctly when entering characters', () => {
      fireEvent.keyDown(mainDiv, { key: 'a' });
      fireEvent.keyDown(mainDiv, { key: 'm' });

      expect(container).toMatchSnapshot();
    });

    it('handles winning correctly', () => {
      fireEvent.keyDown(mainDiv, { key: 'e' });
      fireEvent.keyDown(mainDiv, { key: 's' });
      fireEvent.keyDown(mainDiv, { key: 'n' });
      fireEvent.keyDown(mainDiv, { key: 't' });
      fireEvent.keyDown(mainDiv, { key: 'c' });

      const winSpan = getByText(container, 'You Win!');
      const againButton = getByText(container, 'Try Again');

      expect(winSpan).toBeTruthy();
      expect(againButton).toBeTruthy();
    });

    it('resets the game correctly', () => {
      const againButton = getByText(container, 'Try Again');
      fireEvent.click(againButton);

      expect(container).toMatchSnapshot();
    });

    it('correctly subtracts lives when entering the wrong letter', () => {
      const inputNode = getByLabelText(container, 'Insert the word', {
        selector: 'input',
      });
      const doneButton = getByText(container, 'Done');

      fireEvent.change(inputNode, {
        target: { value: gameSentence },
      });

      fireEvent.click(doneButton);

      const remainingLivesText = num => 'Remaining lives: ' + num;

      const livesSpan = getByText(container, remainingLivesText(7));

      fireEvent.keyDown(mainDiv, { key: 'x' });

      expect(getNodeText(livesSpan)).toEqual(remainingLivesText(6));
    });

    it('correctly shows the game over screen when entering the wrong letter too many times', () => {
      for (let i = 0; i < 7; i++) {
        fireEvent.keyDown(mainDiv, { key: 'x' });
      }

      const gameOverSpan = getByText(container, 'Game Over...');
      const againButton = getByText(container, 'Try Again');

      expect(gameOverSpan).toBeTruthy();
      expect(againButton).toBeTruthy();
    });

    it('resets the game correctly', () => {
      const againButton = getByText(container, 'Try Again');
      fireEvent.click(againButton);

      expect(container).toMatchSnapshot();
    });
  });
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});
