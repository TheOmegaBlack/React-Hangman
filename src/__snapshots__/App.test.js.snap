// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <div
    class="App"
    data-testid="app-root"
    tabindex="1"
  >
    <main
      class="Main"
    >
      <span
        class="Main__title"
      >
        Sacha's Hangman
      </span>
      <label
        class="Main__text"
        for="word-input"
      >
        Insert the word
      </label>
      <input
        autocomplete="off"
        class="Main__input"
        id="word-input"
        type="password"
        value=""
      />
      <button
        class="Main__button"
        type="button"
      >
        Done
      </button>
    </main>
  </div>
</div>
`;

exports[`App when in game scene reacts correctly when entering characters 1`] = `
<div>
  <div
    class="App"
    data-testid="app-root"
    tabindex="1"
  >
    <main
      class="Main"
    >
      <span
        class="Main__title"
      >
        Sacha's Hangman
      </span>
      <span
        class="Main__lives"
      >
        Remaining lives: 
        7
      </span>
      <span
        class="Main__time"
      >
        Remaining time: 
        30
      </span>
      <div
        class="Main__letters"
      >
        <span>
          g 
        </span>
        <span>
          a 
        </span>
        <span>
          m 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          e
        </span>
      </div>
    </main>
  </div>
</div>
`;

exports[`App when in game scene renders correctly after submitting a word 1`] = `
<div>
  <div
    class="App"
    data-testid="app-root"
    tabindex="1"
  >
    <main
      class="Main"
    >
      <span
        class="Main__title"
      >
        Sacha's Hangman
      </span>
      <span
        class="Main__lives"
      >
        Remaining lives: 
        7
      </span>
      <span
        class="Main__time"
      >
        Remaining time: 
        30
      </span>
      <div
        class="Main__letters"
      >
        <span>
          g 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          _ 
        </span>
        <span>
          e
        </span>
      </div>
    </main>
  </div>
</div>
`;

exports[`App when in game scene resets the game correctly 1`] = `
<div>
  <div
    class="App"
    data-testid="app-root"
    tabindex="1"
  >
    <main
      class="Main"
    >
      <span
        class="Main__title"
      >
        Sacha's Hangman
      </span>
      <label
        class="Main__text"
        for="word-input"
      >
        Insert the word
      </label>
      <input
        autocomplete="off"
        class="Main__input"
        id="word-input"
        type="password"
        value=""
      />
      <button
        class="Main__button"
        type="button"
      >
        Done
      </button>
    </main>
  </div>
</div>
`;

exports[`App when in game scene resets the game correctly 2`] = `
<div>
  <div
    class="App"
    data-testid="app-root"
    tabindex="1"
  >
    <main
      class="Main"
    >
      <span
        class="Main__title"
      >
        Sacha's Hangman
      </span>
      <label
        class="Main__text"
        for="word-input"
      >
        Insert the word
      </label>
      <input
        autocomplete="off"
        class="Main__input"
        id="word-input"
        type="password"
        value=""
      />
      <button
        class="Main__button"
        type="button"
      >
        Done
      </button>
    </main>
  </div>
</div>
`;
